### PR TITLE
(WIP) Added database transactions

### PR DESCRIPTION
**Pull request by campbell18**

### CODE REVIEW

File: concord-server/server/src/main/java/com/rabidgremlin/concord/resources/PhrasesResource.java [hunk 1]
```diff
@@ -121,8 +122,7 @@ public Response uploadCsv(@ApiParam(hidden = true) @Auth Caller caller, List<Unl
 		   String phraseId = DigestUtils.md5Hex(unlabelledPhrase.getText());
 		   
 		   // remove any existing votes in case we are reloading data
-		   votesDao.deleteAllVotesForPhrase(phraseId);
-		   phrasesDao.upsert(phraseId, unlabelledPhrase.getText(), false);
+		   uploadDao.deleteExistingVotesAndUploadPhrase(phraseId, unlabelledPhrase);
```
Comment (campbell18) [inline concord-server/server/src/main/java/com/rabidgremlin/concord/resources/PhrasesResource.java]: @rabidgremlin Do you think we should make the entire loop a transaction? Or if we fail keep the ones we managed to upload to the database.

File: concord-server/server/src/main/java/com/rabidgremlin/concord/dao/LabelsDao.java [hunk 2]
```diff
@@ -21,4 +23,18 @@
 	@SqlQuery("SELECT * FROM labels ORDER BY label")
     @RegisterBeanMapper(Label.class)
     List<Label> getLabels();
+
+	@Transaction
+	default void updateLabels(List<Label> labels) {
+		deleteAllLabels();
```
Comment (LaurenceTews) [inline concord-server/server/src/main/java/com/rabidgremlin/concord/dao/LabelsDao.java]: Maybe pull this out to the uploadCsv method, as deleting all labels isn't strictly update functionality.

File: concord-server/server/src/main/java/com/rabidgremlin/concord/resources/PhrasesResource.java (hunk 1 above)
Comment (rabidgremlin) [reply concord-server/server/src/main/java/com/rabidgremlin/concord/resources/PhrasesResource.java]: I think it should be all or nothing. So rollback if a single failure.

File: concord-server/server/src/main/java/com/rabidgremlin/concord/dao/UploadDao.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    VotesDao votesDao();
+
+    @CreateSqlObject
+    PhrasesDao phrasesDao();
+
+    @Transaction
+    default void deleteExistingVotesAndUploadPhrase(String phraseId, UnlabelledPhrase unlabelledPhrase) {
+        votesDao().deleteAllVotesForPhrase(phraseId);
+        phrasesDao().upsert(phraseId, unlabelledPhrase.getText(), false);
+    }
```
Comment (LaurenceTews) [inline concord-server/server/src/main/java/com/rabidgremlin/concord/dao/UploadDao.java]: I think we also want to upsert phrase possible labels (if they exist) into votes here.

File: concord-server/server/src/main/java/com/rabidgremlin/concord/dao/LabelsDao.java (hunk 2 above)
Comment (campbell18) [reply concord-server/server/src/main/java/com/rabidgremlin/concord/dao/LabelsDao.java]: Having it in this class ensures that all the database calls use the same handler - which is needed to ensure they occur in a transaction. 

If we move into the uploadCsv method, we will have to do more lower level handling of the database connection

File: concord-server/server/src/main/java/com/rabidgremlin/concord/dao/UploadDao.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    @CreateSqlObject
+    PhrasesDao phrasesDao();
+
+    @Transaction
+    default void uploadUnlabelledPhrases(List<UnlabelledPhrase> unlabelledPhrases) {
+
+        for(UnlabelledPhrase unlabelledPhrase : unlabelledPhrases)
+        {
+            // skip header
```
Comment (rabidgremlin) [inline concord-server/server/src/main/java/com/rabidgremlin/concord/dao/UploadDao.java]: Pull this out of here. It is a quirk of the .csv file can keep that detail in the resource.

**Pull request merged**